typescript/sdk/src/ism/metadata/arbL2ToL1.ts

```diff
@@ -0,0 +1,169 @@
+import { ArbitrumProvider, L2ToL1MessageReader } from '@arbitrum/sdk';
+import { L2ToL1TransactionEvent } from '@arbitrum/sdk/dist/lib/message/L2ToL1Message.js';
+import { assert } from 'console';
+import { BigNumber, BytesLike, providers, utils } from 'ethers';
+
+import {
+  AbstractMessageIdAuthorizedIsm__factory,
+  ArbSys__factory,
+} from '@hyperlane-xyz/core';
+import {
+  Address,
+  WithAddress,
+  eqAddressEvm,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import { HyperlaneCore } from '../../core/HyperlaneCore.js';
+import { ArbL2ToL1HookConfig } from '../../hook/types.js';
+import { ArbL2ToL1IsmConfig, IsmType } from '../types.js';
+
+import { MetadataBuilder, MetadataContext } from './builder.js';
+
+interface ArbL2ToL1Metadata {
+  proof: BytesLike[];
+  index: BigNumber;
+  l2Sender: Address;
+  to: Address;
+  l2Block: BigNumber;
+  l1Block: BigNumber;
+  l2Timestamp: BigNumber;
+  data: BytesLike;
+}
+
+const ArbSys = ArbSys__factory.createInterface();
+
+export class ArbL2ToL1MetadataBuilder implements MetadataBuilder {
+  constructor(
+    protected readonly core: HyperlaneCore,
+    protected readonly logger = rootLogger.child({
+      module: 'ArbL2ToL1MetadataBuilder',
+    }),
+  ) {}
+
+  async build(
+    context: MetadataContext<
+      WithAddress<ArbL2ToL1IsmConfig>,
+      WithAddress<ArbL2ToL1HookConfig>
+    >,
+  ): Promise<string> {
+    assert(context.ism.type === IsmType.ARB_L2_TO_L1, 'Invalid ISM type');
+    this.logger.debug({ context }, 'Building ArbL2ToL1 metadata');
+
+    // if the executeTransaction call is already successful, we can call with null metadata
+    const ism = AbstractMessageIdAuthorizedIsm__factory.connect(
+      context.ism.address,
+      this.core.multiProvider.getSigner(context.message.parsed.destination),
+    );
+    const verified = await ism.isVerified(context.message.id);
+    if (verified) {
+      this.logger.debug(
+        'Message is already verified, relaying without metadata...',
+      );
+      return '0x';
+    }
+
+    // else build the metadata for outbox.executeTransaction call
+    const matchingL2Tx = context.dispatchTx.logs
+      .filter((log) => eqAddressEvm(log.address, context.hook.arbSys))
+      .map((log) => ArbSys.parseLog(log))
+      .find((log) => {
+        const calldata: string = log.args.data;
+        const messageIdHex = context.message.id.slice(2);
```

Comment:
this is arbitrum message ID?

Reply:
maybe rename for clarity

Reply:
No, it's our messageId

Reply:
Yes, the payload we pass includes our messageId in AbstractMessageIdAuthHook